src/hotspot/share/prims/jvmtiEnv.cpp

```diff
@@ -1644,120 +1644,36 @@ JvmtiEnv::GetFrameCount(JavaThread* java_thread, jint* count_ptr) {
-      if (mh->is_native()) return(JVMTI_ERROR_OPAQUE_FRAME);
-      is_interpreted[frame_count] = vfs.is_interpreted_frame();
-      frame_sp[frame_count] = vfs.frame_id();
-      if (++frame_count > 1) break;
-    }
-    if (frame_count < 2)  {
-      // We haven't found two adjacent non-native Java frames on the top.
-      // There can be two situations here:
-      //  1. There are no more java frames
-      //  2. Two top java frames are separated by non-java native frames
-      if(vframeForNoProcess(java_thread, 1) == NULL) {
-        return JVMTI_ERROR_NO_MORE_FRAMES;
-      } else {
-        // Intervening non-java native or VM frames separate java frames.
-        // Current implementation does not support this. See bug #5031735.
-        // In theory it is possible to pop frames in such cases.
-        return JVMTI_ERROR_OPAQUE_FRAME;
-      }
-    }
-
-    // If any of the top 2 frames is a compiled one, need to deoptimize it
-    EscapeBarrier eb(!is_interpreted[0] || !is_interpreted[1], current_thread, java_thread);
-    for (int i = 0; i < 2; i++) {
-      if (!is_interpreted[i]) {
-        Deoptimization::deoptimize_frame(java_thread, frame_sp[i]);
-        // Eagerly reallocate scalar replaced objects.
-        if (!eb.deoptimize_objects(frame_sp[i])) {
-          // Reallocation of scalar replaced objects failed -> return with error
-          return JVMTI_ERROR_OUT_OF_MEMORY;
-        }
-      }
+  // Eagerly reallocate scalar replaced objects.
+  JavaThread* current_thread = JavaThread::current();
+  EscapeBarrier eb(true, current_thread, java_thread);
+  if (eb.barrier_active()) {
+    if (java_thread->frames_to_pop_failed_realloc() > 0) {
+      // VM is in the process of popping the top frame because it has scalar replaced objects which
+      // could not be reallocated on the heap.
+      // Return JVMTI_ERROR_OUT_OF_MEMORY to avoid interfering with the VM.
+      return JVMTI_ERROR_OUT_OF_MEMORY;
     }
-
-    // Update the thread state to reflect that the top frame is popped
-    // so that cur_stack_depth is maintained properly and all frameIDs
-    // are invalidated.
-    // The current frame will be popped later when the suspended thread
-    // is resumed and right before returning from VM to Java.
-    // (see call_VM_base() in assembler_<cpu>.cpp).
-
-    // It's fine to update the thread state here because no JVMTI events
-    // shall be posted for this PopFrame.
-
-    // It is only safe to perform the direct operation on the current
-    // thread. All other usage needs to use a handshake for safety.
-    {
-      MutexLocker mu(JvmtiThreadState_lock);
-      if (java_thread == JavaThread::current()) {
-        state->update_for_pop_top_frame();
-      } else {
-        UpdateForPopTopFrameClosure op(state);
-        Handshake::execute(&op, java_thread);
-        if (op.result() != JVMTI_ERROR_NONE) {
-          return op.result();
-        }
-      }
+    if (!eb.deoptimize_objects(1)) {
```

Comment:
Oh and why is the depth 1 here, when two frames are deoptimized? Maybe I missed something.

Reply:
@reinrich did I mess something up when merging this in?

Reply:
Depth 1 means top frame and its caller. In UpdateForPopTopFrameClosure::doit() line 1606(?) the 2 top frames are deoptimized. Reallocating objects while a frame pop request is processed does not work if reallocation fails therefore we use an EscapeBarrier to eagerly reallocate objects beforehand.

Reply:
@fisk for PopFrame the top frame needs to be deoptimized (if compiled) to be able to actually remove it when the thread is resumed. Its caller needs to be deoptimized to be able restart the call. For ForceEarlyReturn it is not necessary to restart. The target can return to a compiled caller and continue executing compiled code. So the caller frame is not deoptimized.

@robehn nothing is messed up. Thanks again for doing it.

Reply:
Great!